test/Cardano/AddressDerivationSpec.hs

```diff
@@ -0,0 +1,150 @@
+    , expectFailure
+    , property
+    , vector
+    , (.&&.)
+    , (===)
+    )
+
+import qualified Data.ByteArray as BA
+import qualified Data.ByteString as BS
+
+spec :: Spec
+spec = describe "Checking auxiliary address derivations types" $ do
+    describe "Bounded / Enum relationship" $ do
+        it "The calls Index.succ maxBound should result in a runtime err (hard)"
+            prop_succMaxBoundHardIx
+        it "The calls Index.pred minBound should result in a runtime err (hard)"
+            prop_predMinBoundHardIx
+        it "The calls Index.succ maxBound should result in a runtime err (soft)"
+            prop_succMaxBoundSoftIx
+        it "The calls Index.pred minBound should result in a runtime err (soft)"
+            prop_predMinBoundSoftIx
+
+    describe "Enum Roundtrip" $ do
+        it "Index @'Hardened _" (property prop_roundtripEnumIndexHard)
+        it "Index @'Soft _" (property prop_roundtripEnumIndexSoft)
+
+{-------------------------------------------------------------------------------
+                               Properties
+-------------------------------------------------------------------------------}
+
+prop_succMaxBoundHardIx :: Property
+prop_succMaxBoundHardIx = expectFailure $
+    property $ succ (maxBound @(Index 'Hardened _)) `seq` ()
+
+prop_predMinBoundHardIx :: Property
+prop_predMinBoundHardIx = expectFailure $
+    property $ pred (minBound @(Index 'Hardened _)) `seq` ()
+
+prop_succMaxBoundSoftIx :: Property
+prop_succMaxBoundSoftIx = expectFailure $
+    property $ succ (maxBound @(Index 'Soft _)) `seq` ()
+
+prop_predMinBoundSoftIx :: Property
+prop_predMinBoundSoftIx = expectFailure $
+    property $ pred (minBound @(Index 'Soft _)) `seq` ()
+
+prop_roundtripEnumIndexHard :: Index 'WholeDomain 'AccountK -> Property
+prop_roundtripEnumIndexHard ix =
+    (toEnum . fromEnum) ix === ix .&&. (toEnum . fromEnum . getIndex) ix === ix
+
+prop_roundtripEnumIndexSoft :: Index 'Soft 'AddressK -> Property
+prop_roundtripEnumIndexSoft ix =
+    (toEnum . fromEnum) ix === ix .&&. (toEnum . fromEnum . getIndex) ix === ix
+
+{-------------------------------------------------------------------------------
+                             Arbitrary Instances
+-------------------------------------------------------------------------------}
+
+instance Arbitrary (Index 'Soft 'AddressK) where
+    shrink _ = []
+    arbitrary = arbitraryBoundedEnum
+
+instance Arbitrary (Index 'Hardened 'AccountK) where
+    shrink _ = []
+    arbitrary = arbitraryBoundedEnum
+
+instance Arbitrary (Index 'Hardened 'AddressK) where
+    shrink _ = []
+    arbitrary = arbitraryBoundedEnum
+
+instance Arbitrary (Index 'WholeDomain 'AddressK) where
+    shrink _ = []
+    arbitrary = arbitraryBoundedEnum
+
+instance Arbitrary (Index 'WholeDomain 'AccountK) where
+    shrink _ = []
+    arbitrary = arbitraryBoundedEnum
+
+instance Arbitrary SomeMnemonic where
+    arbitrary = SomeMnemonic <$> genMnemonic @12
```

Comment:
What do you think of moving all the `Arbitrary` instances inside a `Test.Arbitrary` module that every spec file imports? This project will be small enough and tests done in the same context, so it makes sense to share the instances across the various specs here.

Reply:
done in commits : https://github.com/input-output-hk/cardano-addresses/pull/4/commits/fcb6ee2df022d7fbf40f3c4f016311137f50f0b2 https://github.com/input-output-hk/cardano-addresses/pull/4/commits/6a67c795f811ea3a466b69a8bbbbd1024305824d https://github.com/input-output-hk/cardano-addresses/pull/4/commits/8b5fd880422179f83fe3639831bf5a188b85e56a